les/lespay/server/prioritypool.go

```diff
@@ -0,0 +1,502 @@
+		pp.activeCount++
+	}
+	if cap == 0 {
+		pp.activeCount--
+	}
+	n.capacity = cap
+}
+
+// enforceLimits enforces active node count and total capacity limits. It returns the
+// lowest active node priority. Note that this function is performed on the temporary
+// internal state.
+func (pp *PriorityPool) enforceLimits() int64 {
+	if pp.activeCap <= pp.maxCap && pp.activeCount <= pp.maxCount {
+		return math.MinInt64
+	}
+	var maxActivePriority int64
+	pp.activeQueue.MultiPop(func(data interface{}, priority int64) bool {
+		c := data.(*ppNodeInfo)
+		pp.markForChange(c)
+		maxActivePriority = priority
+		if c.capacity == pp.minCap {
+			pp.setCapacity(c, 0)
+		} else {
+			sub := c.capacity / pp.capacityStepDiv
+			if c.capacity-sub < pp.minCap {
+				sub = c.capacity - pp.minCap
+			}
+			pp.setCapacity(c, c.capacity-sub)
+			pp.activeQueue.Push(c)
+		}
+		return pp.activeCap > pp.maxCap || pp.activeCount > pp.maxCount
+	})
+	return invertPriority(maxActivePriority)
+}
+
+// finalizeChanges either commits or reverts temporary changes. The necessary capacity
+// field and according flag updates are not performed here but returned in a list because
+// they should be performed while the mutex is not held.
+func (pp *PriorityPool) finalizeChanges(commit bool) (updates []capUpdate) {
+	for _, c := range pp.changed {
+		// always remove and push back in order to update biased/forced priority
+		pp.activeQueue.Remove(c.activeIndex)
+		pp.inactiveQueue.Remove(c.inactiveIndex)
+		c.bias = 0
+		c.forced = false
+		c.changed = false
+		if !commit {
+			pp.setCapacity(c, c.origCap)
+		}
+		if c.connected {
+			if c.capacity != 0 {
+				pp.activeQueue.Push(c)
+			} else {
+				pp.inactiveQueue.Push(c, pp.inactivePriority(c))
+			}
+			if c.capacity != c.origCap && commit {
+				updates = append(updates, capUpdate{c.node, c.origCap, c.capacity})
+			}
+		}
+		c.origCap = 0
+	}
+	pp.changed = nil
+	return
+}
+
+// capUpdate describes a CapacityField and ActiveFlag/InactiveFlag update
+type capUpdate struct {
+	node           *enode.Node
+	oldCap, newCap uint64
+}
+
+// updateFlags performs CapacityField and ActiveFlag/InactiveFlag updates while the
+// pool mutex is not held
+// Note: this function should run inside a NodeStateMachine operation
+func (pp *PriorityPool) updateFlags(updates []capUpdate) {
+	for _, f := range updates {
+		pp.ns.SetFieldSub(f.node, pp.CapacityField, f.newCap)
+		if f.oldCap == 0 {
+			pp.ns.SetStateSub(f.node, pp.ActiveFlag, pp.InactiveFlag, 0)
+		}
+		if f.newCap == 0 {
+			pp.ns.SetStateSub(f.node, pp.InactiveFlag, pp.ActiveFlag, 0)
+		}
+	}
+}
+
+// tryActivate tries to activate inactive nodes if possible
+func (pp *PriorityPool) tryActivate() []capUpdate {
+	var commit bool
+	for pp.inactiveQueue.Size() > 0 {
+		c := pp.inactiveQueue.PopItem().(*ppNodeInfo)
+		pp.markForChange(c)
+		pp.setCapacity(c, pp.minCap)
+		c.bias = pp.activeBias
+		pp.activeQueue.Push(c)
+		pp.enforceLimits()
+		if c.capacity > 0 {
+			commit = true
```

Comment:
Is it possible that the first element in the `inactiveQueue` is activated which sets the `commit` as true.
While this loop will try to activate the next one but it's failed.

In this case, the `commit` is still true so that we will try to finalize the changes.

Reply:
If the second client cannot be activated then `pp.enforceLimits()` will restore a consistent state where the second inactive client is set back to inactive state. So the finalized state will be correct with only the first inactive client activated.

Reply:
> will restore a consistent state where the second inactive client is set back to inactive state

It's true that the status of the second client will be reset. But during the activation of the second client, we will also do some changes to the queue, e.g. decrease capacity of active client, or set the cap to 0. **we don't revert these changes**

Reply:
Have you actually seen incorrect behavior? To me it still looks good. `enforceLimits` only removes or reduces the capacity of the lowest priority active clients while the total capacity is too high. So if the second inactive client gets into the active queue and it does not fit there, it is the one that is going to be removed again by `enforceLimits` and the other active clients are left untouched. If another active client is kicked out because it has a lower priority now than the recently activated then it is fine too. Sometimes inactive clients can kick out active ones if priorities have changed. The important point is that `enforceLimits` never kicks an active client that should stay there.

Reply:
I see, I misunderstand previously. I thought we should guarantee the atomicity of the operation.
However, you still want to ensure the total capacity is under the limitation.

Reply:
It is atomic because everything happens under `pp.lock` and the limits are always enforced before the lock is released. We try activating inactive nodes (which can raise the total cap over the limit) and then call `enforceLimits` which fixes this either by kicking another client and keeping the recently activated one active (in which case its capacity will be non-zero after the enforcing) or finally kicks the last activated one (in which case we stop trying and return with a consistent state).